File: additional exercise series/series_08 [Python]/D/Dumas Cipher.py
def cipherkey(keytext: str):
    keytext = [char.upper() for char in keytext if char not in (' ', '\t', '\n')]
    key = dict()
    for index, letter in enumerate(keytext):
        if letter not in key:
            key[letter] = list()
        key[letter].append(index + 1)
    return key


def encode(plaintext: str, keytext: str):
    key = cipherkey(keytext)
    index_count = dict.fromkeys(key, 0)
    encoded = list()
    for letter in plaintext.upper():
        key_letter = key[letter]
        encoded.append(key_letter[index_count[letter] % len(key_letter)])
        index_count[letter] += 1
    return encoded


keytext = 'Lost time is never found again.'
print(encode('nondeterminativeness', keytext))
print(encode('interdenominationalism', keytext))
print(encode('gastroenteroanastomosis', keytext))

File: exercise series (2019–2020)/week_08 [JavaScript]/Dumas cipher.js
function codeersleutel(sleuteltekst) {
    sleuteltekst = sleuteltekst.replace(/\s+/g, "").toUpperCase();
    let sleutel = {};
    for (const [index, letter] of sleuteltekst.split("").entries()) {
        if (!sleutel.hasOwnProperty(letter)) sleutel[letter] = [];
        sleutel[letter].push(index + 1);
    }
    return sleutel;
}

function codeer(tekst, sleuteltekst) {
    const [encoded, indices, sleutel] = [[], {}, codeersleutel(sleuteltekst)];
    for (const letter in sleutel) indices[letter] = 0;
    for (const letter of tekst.toUpperCase()) {
        const sleutel_letter = sleutel[letter];
        encoded.push(sleutel_letter[indices[letter] % sleutel_letter.length]);
        indices[letter] += 1;
    }
    return encoded;
}

const sleuteltekst = 'Lost time is never found again.';
console.log(codeer('nondeterminativeness', sleuteltekst));
console.log(codeer('interdenominationalism', sleuteltekst));
console.log(codeer('gastroenteroanastomosis', sleuteltekst));